### DISCUSSION TITLE
ArduinoOTA - ESP32 via Ethernet port

### DISCUSSION THREAD

**Discussion by OI-Pietro** [JAndrassy/ArduinoOTA#83] [General]
Dear Jandrassy, Great Work!
I've few ours on my shoulders in Arduino programming so this lack of experience doesn't help me. My goal is try the OTA of ESP32 module by downloading a bin located on a server web by using ethernet port. With bundled ArduinoOTA library I've already updated by Wifi the ESP32 but now i try by Ethernet with the help of your ArduinoOTA library and ArduinoHttpClient. The Ethernet module used is WIZ610 (chip W6100) similar to W5500, this requires the Ethernet Library 2.0 modified by WIZnet.  
Compiling your ino OTASketchDownload.ino with Arduino IDE, occurs the following errors:
>>>

_In file included from C:\Users\ing2\AppData\Local\Temp\arduino_modified_sketch_301356\sketch_apr20b.ino:17:0:
...esp32\1.0.5\libraries\ArduinoOTA\src/ArduinoOTA.h:123:22: error: 'WiFiServer' was not declared in this scope
 ArduinoOTAMdnsClass <WiFiServer, WiFiClient, WiFiUDP> ArduinoOTA;
                      ^
...esp32\1.0.5\libraries\ArduinoOTA\src/ArduinoOTA.h:123:34: error: 'WiFiClient' was not declared in this scope
 ArduinoOTAMdnsClass <WiFiServer, WiFiClient, WiFiUDP> ArduinoOTA;
                                  ^
...esp32\1.0.5\libraries\ArduinoOTA\src/ArduinoOTA.h:123:53: error: template argument 1 is invalid
 ArduinoOTAMdnsClass <WiFiServer, WiFiClient, WiFiUDP> ArduinoOTA;_

<<<

Once seen the error with wifi istances, i tried to ad a "  #define ethernet_h  and #define ESP32. Now the error is:

>>>
_...esp32\1.0.5\libraries\ArduinoOTA\src/ArduinoOTA.h:112:23: error: 'EthernetServer' was not declared in this scope
 ArduinoOTAMdnsClass  <EthernetServer, EthernetClient, EthernetUDP>   ArduinoOTA;
                       ^
...esp32\1.0.5\libraries\ArduinoOTA\src/ArduinoOTA.h:112:39: error: 'EthernetClient' was not declared in this scope
 ArduinoOTAMdnsClass  <EthernetServer, EthernetClient, EthernetUDP>   ArduinoOTA;
                                       ^
...esp32\1.0.5\libraries\ArduinoOTA\src/ArduinoOTA.h:112:55: error: 'EthernetUDP' was not declared in this scope
 ArduinoOTAMdnsClass  <EthernetServer, EthernetClient, EthernetUDP>   ArduinoOTA;
                                                       ^
...esp32\1.0.5\libraries\ArduinoOTA\src/ArduinoOTA.h:112:66: error: template argument 1 is invalid
 ArduinoOTAMdnsClass  <EthernetServer, EthernetClient, EthernetUDP>   ArduinoOTA;
                                                                  ^
...esp32\1.0.5\libraries\ArduinoOTA\src/ArduinoOTA.h:112:66: error: template argument 2 is invalid
---esp32\1.0.5\libraries\ArduinoOTA\src/ArduinoOTA.h:112:66: error: template argument 3 is invalid_
<<<
Tnks

**Comment by JAndrassy** [comment]
the first problem with ESP32 core even with the standard Ethernet library is that the core has broken  compatibility of Arduino API Server class. They changed it and don't want to change it back so any library which uses Server is incompatible.

You can make it compatible by adding `virtual void begin(uint16_t port) {};` to EthernetServer class

**Reply by OI-Pietro** [reply]
tnks for your support; By observing the portion of Ethernet library :
<<<<<
uint16_t EthernetServer::server_port[MAX_SOCK_NUM];
void EthernetServer::begin()
{
	uint8_t sockindex = Ethernet.socketBegin(SnMR::TCP, _port);
	if (sockindex < MAX_SOCK_NUM) {
		if (Ethernet.socketListen(sockindex)) {
			server_port[sockindex] = _port;
		} else {
			Ethernet.socketDisconnect(sockindex);
		}
	}
}  
<<<<<<
do you means that it is enough to modify it as the follow? :
<<<<<
uint16_t EthernetServer::server_port[MAX_SOCK_NUM];
_**virtual**_ void EthernetServer::begin()
{
	uint8_t sockindex = Ethernet.socketBegin(SnMR::TCP, _port);
	if (sockindex < MAX_SOCK_NUM) {
		if (Ethernet.socketListen(sockindex)) {
			server_port[sockindex] = _port;
		} else {
			Ethernet.socketDisconnect(sockindex);
		}
	}
}
<<<<<

**Reply by JAndrassy** [reply]
only add the one line from my comment to class EthernetServer in .h

**Reply by OI-Pietro** [reply]
excuse the delay but i was full with other project. I re-try today your suggestion (as in the attached file i add your line  " virtual void begin(uint16_t port) {}; " to EthernetServer class in the Ethernet.h. The result is the same: each argument of ArduinoOTAMdnsClass seem to be not declared/invalid:
<<<<<<<<<<<<<<
In file included from C:\Users\ing\Documents\Arduino\sketch_jul13a\sketch_jul13a.ino:17:0:
C:\Users\ing\AppData\Local\Arduino15\packages\esp32\hardware\esp32\1.0.4\libraries\ArduinoOTA-master\src/ArduinoOTA.h:112:23: error: 'EthernetServer' was not declared in this scope
 ArduinoOTAMdnsClass  <EthernetServer, EthernetClient, EthernetUDP>   ArduinoOTA;
                       ^
C:\Users\ing\AppData\Local\Arduino15\packages\esp32\hardware\esp32\1.0.4\libraries\ArduinoOTA-master\src/ArduinoOTA.h:112:39: error: 'EthernetClient' was not declared in this scope
 ArduinoOTAMdnsClass  <EthernetServer, EthernetClient, EthernetUDP>   ArduinoOTA;
                                       ^
C:\Users\ing\AppData\Local\Arduino15\packages\esp32\hardware\esp32\1.0.4\libraries\ArduinoOTA-master\src/ArduinoOTA.h:112:55: error: 'EthernetUDP' was not declared in this scope
 ArduinoOTAMdnsClass  <EthernetServer, EthernetClient, EthernetUDP>   ArduinoOTA;
                                                       ^
C:\Users\ing\AppData\Local\Arduino15\packages\esp32\hardware\esp32\1.0.4\libraries\ArduinoOTA-master\src/ArduinoOTA.h:112:66: error: template argument 1 is invalid
 ArduinoOTAMdnsClass  <EthernetServer, EthernetClient, EthernetUDP>   ArduinoOTA;
                                                                  ^
C:\Users\ing\AppData\Local\Arduino15\packages\esp32\hardware\esp32\1.0.4\libraries\ArduinoOTA-master\src/ArduinoOTA.h:112:66: error: template argument 2 is invalid
C:\Users\ing\AppData\Local\Arduino15\packages\esp32\hardware\esp32\1.0.4\libraries\ArduinoOTA-master\src/ArduinoOTA.h:112:66: error: template argument 3 is invalid
exit status 1
>>>>>>>>>>

Many thanks again.......


![Cattura](https://user-images.githubusercontent.com/67467327/125962841-f451ea80-ec22-4315-a5d3-04644803f2c6.PNG)

**Comment by JAndrassy** [comment]
did you include Ethernet.h before ArduinoOTA.h?

**Comment by OI-Pietro** [comment]
No….. now I’ve inverted the including order and compiled successfully!!!

I’m going to start a server with a bin file and test ……. I hope to do not have further problems…J

 

Many and many thanks for your work and precious suggestions. 

Have good day!

Pietro

 

 

Da: Juraj Andrássy ***@***.*** 
Inviato: venerdì 16 luglio 2021 16:57
A: jandrassy/ArduinoOTA
Cc: OI-Pietro; Author
Oggetto: Re: [jandrassy/ArduinoOTA] ArduinoOTA - ESP32 via Ethernet port (#83)

 

did you include Ethernet.h before ArduinoOTA.h?
